File: src/test/typescript/accounting-pattern/AcceptanceTest.ts
import { Customer } from "../../main/typescript/accounting-pattern/Customer";
import { ServiceAgreement } from "../../main/typescript/accounting-pattern/ServiceAgreement";
import { EventType } from "../../main/typescript/accounting-pattern/EventType";
import { MultiplyByRatePostingRule } from "../../main/typescript/accounting-pattern/MultiplyByRatePostingRule";
import { AnnualServiceFeeRule } from "./AnnualServiceFeeRule";
import { EntryType } from "../../main/typescript/accounting-pattern/EntryType";
import { Quantity } from "../../main/typescript/accounting-pattern/Quantity";

describe("Acceptance Test", () => {
    it("should pass the acceptance test", () => {
        const effectiveDate = new Date(1999, 9, 1);
        const customer = new Customer("Acme Coffee Makers",
            new ServiceAgreement(10)
                .addPostingRule(EventType.USAGE, new MultiplyByRatePostingRule(EntryType.BASE_USAGE), effectiveDate)
                .addPostingRule(EventType.SERVICE, new AnnualServiceFeeRule(EntryType.SERVICE, 100), effectiveDate)
                .addPostingRule(EventType.SERVICE, new AnnualServiceFeeRule(EntryType.SERVICE, 1000), new Date(2001, 9, 1))
        );

        customer.post(EventType.USAGE, new Date(1999, 11, 1), new Quantity(50));
        customer.post(EventType.SERVICE, new Date(2000, 9, 1), new Quantity(1));

        expect(customer.balance(new Date(2020, 9, 1)).getAmount()).toBe(600);

        customer.post(EventType.USAGE, new Date(1999, 11, 1), new Quantity(50));

        expect(customer.balance(new Date(2020, 9, 1)).getAmount()).toBe(1100);

        customer.post(EventType.SERVICE, new Date(2001, 10, 1), new Quantity(1));
        expect(customer.balance(new Date(2020, 9, 1)).getAmount()).toBe(2100);
    });
});

File: src/test/java/accounting/patterns/AcceptanceTest.java
package accounting.patterns;

import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.time.LocalDate;

import static org.junit.jupiter.api.Assertions.assertEquals;

class AcceptanceTest {
    @Test
    void balanceAccumulatesEveryPostedEvent() {
        LocalDate effectiveDate = LocalDate.of(1999, 10, 1);
        LocalDate balanceDate = LocalDate.of(2020, 10, 1);
        var customer = new Customer("Acme Coffee Makers",
                new ServiceAgreement(new BigDecimal(10))
                        .addPostingRule(EventType.USAGE, new MultiplyByRatePostingRule(EntryType.BASE_USAGE), effectiveDate)
                        .addPostingRule(EventType.SERVICE, new AnnualServiceFeeRule(EntryType.SERVICE, 100), effectiveDate)
                        .addPostingRule(EventType.SERVICE, new AnnualServiceFeeRule(EntryType.SERVICE, 1000), effectiveDate.plusYears(2))
        );

        customer.post(EventType.USAGE, effectiveDate.plusMonths(2), new Quantity(50));
        customer.post(EventType.SERVICE, effectiveDate.plusMonths(12), new Quantity(1));

        assertEquals(MonetaryAmount.of("USD", 600), customer.balance(balanceDate));

        customer.post(EventType.USAGE, effectiveDate.plusMonths(2), new Quantity(50));

        assertEquals(MonetaryAmount.of("USD", 1100), customer.balance(balanceDate));

        // Past the second effective date, the later service fee rule applies.
        customer.post(EventType.SERVICE, effectiveDate.plusMonths(25), new Quantity(1));

        assertEquals(MonetaryAmount.of("USD", 2100), customer.balance(balanceDate));
    }
}
